src/OrchardCore.Modules/OrchardCore.Media/Services/ChunkFileUploadService.cs

```diff
@@ -0,0 +1,262 @@
+    private const String UploadIdFormKey = "__chunkedFileUploadId";
+    private const String TempFolderPrefix = "ChunkedFileUploads";
+    private static readonly SemaphoreSlim _purgeLock = new(1, 1);
+    private readonly IOptions<MediaOptions> _options;
+    private readonly ILogger _logger;
+    private readonly IClock _clock;
+
+    public ChunkFileUploadService(
+        IClock clock,
+        ILogger<ChunkFileUploadService> logger,
+        IOptions<MediaOptions> options)
+    {
+        _clock = clock;
+        _logger = logger;
+        _options = options;
+    }
+
+    public async Task<IActionResult> ProcessRequestAsync(
+        HttpRequest request,
+        Func<Guid, IFormFile, ContentRangeHeaderValue, Task<IActionResult>> chunkAsync,
+        Func<IEnumerable<IFormFile>, Task<IActionResult>> completedAsync)
+    {
+        var contentRangeHeader = request.Headers.ContentRange;
+
+        if (_options.Value.MaxUploadChunkSize <= 0
+            || contentRangeHeader.Count is 0
+            || !request.Form.TryGetValue(UploadIdFormKey, out var uploadIdValue))
+        {
+            return await completedAsync(request.Form.Files);
+        }
+
+        if (request.Form.Files.Count != 1
+            || !ContentRangeHeaderValue.TryParse(contentRangeHeader, out var contentRange)
+            || !Guid.TryParse(uploadIdValue, out var uploadId)
+            || !contentRange.HasLength
+            || !contentRange.HasRange
+            || contentRange.Length > _options.Value.MaxFileSize
+            || contentRange.To - contentRange.From > _options.Value.MaxUploadChunkSize)
+        {
+            return new BadRequestResult();
+        }
+
+        var formFile = request.Form.Files.First();
+        using (var fileStream = GetOrCreateTemporaryFile(
+            uploadId,
+            formFile,
+            contentRange.Length.Value))
+        {
+            fileStream.Seek(contentRange.From.Value, SeekOrigin.Begin);
+            await formFile.CopyToAsync(fileStream, request.HttpContext.RequestAborted);
+        }
+
+        return (contentRange.To.Value + 1) >= contentRange.Length.Value
+            ? await CompleteUploadAsync(uploadId, formFile, completedAsync)
+            : await chunkAsync(uploadId, formFile, contentRange);
+    }
+
+    private async Task<IActionResult> CompleteUploadAsync(
+        Guid uploadId,
+        IFormFile formFile,
+        Func<IEnumerable<IFormFile>, Task<IActionResult>> completedAsync)
+    {
+        var chunkedFormFile = GetTemporaryFileForRead(uploadId, formFile);
+
+        try
+        {
+            return await completedAsync(new[] { chunkedFormFile });
+        }
+        finally
+        {
+            chunkedFormFile.Dispose();
+            DeleteTemporaryFile(uploadId, formFile);
+            PurgeTempDirectory();
+        }
+    }
+
+    public void PurgeTempDirectory()
+    {
+        if (_purgeLock.CurrentCount == 0)
+        {
+            return;
+        }
+
+        _purgeLock.Wait();
```

Comment:
So maybe we could remove the usage of the `SemaphoreSlim` if each tenant only remove their cached file and if `PurgeTempDirectory()` is only called by the background task.